### PR TITLE
udpa: plumb udpa:// resource names/locators.

### PR DESCRIPTION
**Pull request by htuch**
These are currently set as [#not-implemented-hide:]. There should be enough plumbing here to start
implementing the first roadmap steps in #11264, i.e. some simple delta discovery flows with core
LDS/CDS/RDS/EDS/SDS.

I've punted on VHDS and SRDS until later, as these will require special case mapping of their
resource keys into context parameters.

Risk level: Low (Unused API changes).
Testing: Built docs and proto_format.sh.

Part of #11264.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 1]
```diff
@@ -72,11 +76,19 @@ message Bootstrap {
     // :ref:`LDS <arch_overview_dynamic_config_lds>` configuration source.
     core.v3.ConfigSource lds_config = 1;
 
+    // Resource locator for listener collection.
+    // [#not-implemented-hide:]
+    udpa.core.v1.ResourceLocator lds_resources_locator = 5;
```
Comment (mattklein123) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:81]: Is the plan to deprecate the parallel config source fields when these are implemented?

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 2]
```diff
@@ -186,6 +198,29 @@ message Bootstrap {
...
+  // Configuration sources that will participate in
+  // *udpa.core.v1.ResourceLocator* authority resolution. The algorithm is as
+  // follows:
+  // 1. The authority field is taken from the *udpa.core.v1.ResourceLocator*, call
+  //    this *resource_authority*.
+  // 2. *resource_authority* is compared against the authorities in any peer
+  //    *ConfigSource*. The peer *ConfigSource* is the configuration source
+  //    message which would have been used unconditionally for resolution
+  //    with opaque resource names. If there is a match with an authority, the
+  //    peer *ConfigSource* message is used.
```
Comment (mattklein123) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:211]: OK I see, this answers my question above. I think in the final version this doc should be somewhere else and linked to from everywhere this is relevant? Otherwise this is pretty confusing?

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 3]
```diff
@@ -186,6 +198,29 @@ message Bootstrap {
...
+  //    to match wins.
+  // 4. As a fallback, if no configuration source matches, then
+  //    *default_configuration_source* is used.
+  // [#not-implemented-hide:]
+  repeated core.v3.ConfigSource config_sources = 22;
+
+  // Default configuration source for *udpa.core.v1.ResourceLocator* if all
+  // other resolution fails.
+  // [#not-implemented-hide:]
+  core.v3.ConfigSource default_config_source = 23;
```
Comment (mattklein123) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:223]: Would this end up being required in v4? If not required, can we ultimately document what the behavior is if there is no matching config source for an authority?

File: api/envoy/config/cluster/v3/cluster.proto [hunk 4]
```diff
@@ -178,7 +188,12 @@ message Cluster {
     // Optional alternative to cluster name to present to EDS. This does not
     // have the same restrictions as cluster name, i.e. it may be arbitrary
     // length.
-    string service_name = 2;
+    string service_name = 2 [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
+
+    // Resource locator for EDS. This is mutually exclusive to *service_name*.
+    // [#not-implemented-hide:]
+    udpa.core.v1.ResourceLocator eds_resource_locator = 3
+        [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
```
Comment (markdroth) [inline api/envoy/config/cluster/v3/cluster.proto:196]: Don't we want this field to replace both the `service_name` field and the `eds_config` field?  In other words, I think we want either this new field or *both* `eds_config` and `service_name`.

Maybe the right thing to do here is to leave the `EdsClusterConfig` message alone and move this new field into a top-level field in the `Cluster` message.  Then the `oneof` can contain this new field and the `EdsClusterConfig` field.

File: api/envoy/extensions/common/tap/v3/common.proto [hunk 5]
```diff
@@ -30,7 +33,12 @@ message CommonExtensionConfig {
     config.core.v3.ConfigSource config_source = 1 [(validate.rules).message = {required: true}];
 
     // Tap config to request from XDS server.
-    string name = 2 [(validate.rules).string = {min_bytes: 1}];
+    string name = 2 [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
+
+    // Resource locator for TAP. This is mutually exclusive to *name*.
+    // [#not-implemented-hide:]
+    udpa.core.v1.ResourceLocator tap_resource_locator = 3
```
Comment (markdroth) [inline api/envoy/extensions/common/tap/v3/common.proto:40]: Similar question here: shouldn't this new field also replace use of the existing `config_source` field?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 6]
```diff
@@ -602,7 +605,13 @@ message Rds {
   // API. This allows an Envoy configuration with multiple HTTP listeners (and
   // associated HTTP connection manager filters) to use different route
   // configurations.
-  string route_config_name = 2 [(validate.rules).string = {min_bytes: 1}];
+  string route_config_name = 2
+      [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
+
+  // Resource locator for RDS. This is mutually exclusive to *route_config_name*.
+  // [#not-implemented-hide:]
+  udpa.core.v1.ResourceLocator rds_resource_locator = 3
```
Comment (markdroth) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:613]: Same question here w.r.t. replacing the `config_source` field.

File: api/envoy/extensions/transport_sockets/tls/v3/secret.proto [hunk 7]
```diff
@@ -30,7 +33,12 @@ message SdsSecretConfig {
   // Name (FQDN, UUID, SPKI, SHA256, etc.) by which the secret can be uniquely referred to.
   // When both name and config are specified, then secret can be fetched and/or reloaded via
   // SDS. When only name is specified, then secret will be loaded from static resources.
-  string name = 1;
+  string name = 1 [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
+
+  // Resource locator for SDS. This is mutually exclusive to *name*.
+  // [#not-implemented-hide:]
+  udpa.core.v1.ResourceLocator sds_resource_locator = 3
```
Comment (markdroth) [inline api/envoy/extensions/transport_sockets/tls/v3/secret.proto:40]: Same question here w.r.t. replacing the existing `config_source` field.

File: api/envoy/service/discovery/v3/discovery.proto [hunk 8]
```diff
@@ -174,9 +179,19 @@ message DeltaDiscoveryRequest {
   // A list of Resource names to add to the list of tracked resources.
   repeated string resource_names_subscribe = 3;
 
+  // As with *resource_names_subscribe* but used when subscribing to resources indicated
+  // by a *udpa.core.v1.ResourceLocator*.
+  // [#not-implemented-hide:]
+  repeated udpa.core.v1.ResourceName udpa_resources_subscribe = 8;
```
Comment (markdroth) [inline api/envoy/service/discovery/v3/discovery.proto:185]: Shouldn't this be a `ResourceLocator`, not a `ResourceName`?

File: api/envoy/service/discovery/v3/discovery.proto [hunk 9]
```diff
@@ -174,9 +179,19 @@ message DeltaDiscoveryRequest {
...
+
   // A list of Resource names to remove from the list of tracked resources.
   repeated string resource_names_unsubscribe = 4;
 
+  // As with *resource_names_unsubscribe* but used when unsubscribing to resources indicated by a
+  // *udpa.core.v1.ResourceLocator*.
+  // [#not-implemented-hide:]
+  repeated udpa.core.v1.ResourceName udpa_resources_unsubscribe = 9;
```
Comment (markdroth) [inline api/envoy/service/discovery/v3/discovery.proto:193]: Same here.

File: api/envoy/service/discovery/v3/discovery.proto [hunk 10]
```diff
@@ -215,22 +230,34 @@ message DeltaDiscoveryResponse {
...
   // The resource's name, to distinguish it from others of the same type of resource.
-  string name = 3;
+  string name = 3 [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
+
+  // Used instead of *name* when a resource with a *udpa.core.v1.ResourceName* is delivered.
+  udpa.core.v1.ResourceName udpa_resource_name = 5
+      [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
 
   // The aliases are a list of other names that this resource can go by.
   repeated string aliases = 4;
```
Comment (markdroth) [inline api/envoy/service/discovery/v3/discovery.proto:263]: We're going to deprecate this, right?  Are we going to do that in a separate PR?

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 1 above)
Comment (htuch) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:81]: I think it might be fine to keep these, as per the resolution algorithm.

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 3 above)
Comment (htuch) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:223]: I think if none is specified, we can say resolution fails. I think ultimately we might want to consider some kind of fall-back DNS resolution as default, but that will require some thought.

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:211]: Yeah, I plan on doing a major documentation round before we consider `udpa://` ready for use.

File: api/envoy/config/cluster/v3/cluster.proto (hunk 4 above)
Comment (htuch) [reply api/envoy/config/cluster/v3/cluster.proto:196]: So, as per the resolution algorithm in the bootstrap, I think it's fine to have "peer config sources". This get checked first, then the bootstrap. It provides a more gradual path to adoption for folks who don't already provide config sources for everything in bootstrap.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 8 above)
Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto:185]: In the design document it is a resource locator, but I'm wondering if that was a mistake. The client will only ever ask for a named resource, directives are resolved locally.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 10 above)
Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto:263]: Yes, separate PR. I've avoided adding any deprecations as the new alternatives are not ready yet, so wanted to avoid the "deprecated but not ready yet" anti-pattern ;-D.

File: api/envoy/service/discovery/v3/discovery.proto (hunk 8 above)
Comment (markdroth) [reply api/envoy/service/discovery/v3/discovery.proto:185]: I don't think that was a mistake.  The whole point of the division between `ResourceLocator` and `ResourceName` is that the former is to be used for matching semantics in requests and the latter is to be used to name concrete resources in responses.  Glob patterns need to be expressed as a `ResourceLocator`, not a `ResourceName`.  And if in the future we add more flexible matching for context params, that would be encoded in `ResourceLocator` as well.

It's true that `ResourceLocator` includes directives, but I think that's okay.  We went back and forth on this in the design, and I think where we wound up was that we would just not use the directives field of `ResourceLocator` in the context of `DeltaDiscoveryRequest`.  But I would also be okay with the other approach we had discussed of splitting up `ResourceLocator` into a base message without directives and a wrapper message that includes the base message and adds directives.  We could then use the base message here and the wrapper message in resources we send to the client.

Comment (htuch) [reply api/envoy/service/discovery/v3/discovery.proto]: Right, thanks for the reminder. I've updated to make these locators.

**Comment by htuch** [comment]
@mattklein123 any additional comments or changes you'd like to see happen?
